Feature/670 display watershed size

maxdiebold-erg opened the pull request:
## Related Issues:
* [HMW-670](https://jira.epa.gov/browse/HMW-670)

## Main Changes:
* Added popups for the Watersheds layer and the Boundaries layer.
  * These popups are identical: if both are visible and clicked, only one will be shown.
* Removed the default handling of popups universally.
  * Previously, this was only done on the Tribal page and the Monitoring tab, but the Watersheds layer is available everywhere, and we'd like the watershed popups to always be last. To my knowledge, this only affected the Protected Areas layer (a MapImageLayer) popups, which need to fetch from the server before showing. I changed this to instead fetch the features manually if the layer is visible. This also fixes the popups in the two locations mentioned previously. I wasn't able to recreate the Esri cursor spinner, though, so the default browser wait cursor is used instead.
* Added the watershed area (in acres and square kilometers) to the watershed summary information to the right of the pin icon above the tabs on the Community page.
* Updated the Upstream Watershed popup to use two points of precision for consistency with the other watersheds popup.

## Steps To Test:
1. Go to http://localhost:3000/community/dc/overview.
2. Click outside the Boundaries layer. The "Change to This Location?" popup should appear.
3. Toggle on the visibility of the Watersheds layer, then click outside the Boundaries layer again. A watershed popup should show with a "Change to this location?" button within.
4. With the Watersheds layer still visible, click inside the boundaries of the current HUC. Only one watershed popup should show, and no "Change to this location?" message should be displayed.
5. Toggle on the visibility of the Protected Areas layer, then click on of its features. The cursor should change to a wait cursor inside the map container, then a popup should show with one or more Protected Areas features.
6. Repeat steps 1-5 for other pages.

maxdiebold-erg commented:
I realized the conditional that checks whether to fetch Protected Areas features doesn't account for the scale, so I'm adding that check.

cschwinderg commented on app/client/src/components/pages/Community.Routes.CommunityTabs.js:
```diff
@@ -451,9 +453,15 @@ function CommunityTabs() {
+          )}
+          {watershed.areasqkm && watershed.areaacres && (
+            <p css={watershedStyles}>
+              <span>SIZE:</span> {formatNumber(watershed.areaacres)} acres /{' '}
+              {formatNumber(watershed.areasqkm, 2)} km<sup>2</sup>
```
I think we should be consistent with the area. Either we should show both acreas and sqkm in popups or this should be updated to only show sqkm.

maxdiebold-erg replied:
I agree, @coobr01 suggested the former for the summary info, so I'll go ahead and change the popup to match.

cschwinderg commented on app/client/src/components/pages/Community.Tabs.Protect.js:
```diff
@@ -571,7 +571,7 @@ function Protect() {
                           <p>
                             No Watershed Health Score data available for the{' '}
-                            <em>{watershed}</em> watershed.
+                            <em>{watershed.name}</em> watershed.
```
Line `586` of this file needs to be changed from `value: watershed` to `value: watershed.name`.

maxdiebold-erg replied:
Good catch!

maxdiebold-erg commented:
I added the scale check, and I also updated the `isInScale` function: since `minScale` and `maxScale` can be set on group/parent layers and sublayers, I think it makes more sense for that function to not recurse. It doesn't look like the recursive behavior was being utilized anymore, either, since it's only used in the `MapLegend` component, otherwise.

cschwinderg reviewed (approved):
Works well!

I left one suggestion below. I also wonder if we should display the area in the "Change to this location?" popup when the watersheds layer is off.

cschwinderg reviewed (changes requested):
I meant to set this to "Request changes"

maxdiebold-erg commented:
@cschwinderg I see what you mean about showing the area in the change location popup, but it seems like it could be extraneous information when there are is no feature visible with the associated area. Maybe bring it up during the meeting tomorrow?

cschwinderg commented:
That sounds good. I figured I would point it out, since we do display the watershed name and huc there.

cschwinderg commented on app/client/src/components/shared/WaterbodyInfo.tsx:
```diff
@@ -928,11 +928,16 @@ function WaterbodyInfo({
       {labelValue(
         'Area',
-        attributes.areasqkm && `${formatNumber(attributes.areasqkm)} sq. km.`,
+        attributes.areasqkm &&
+          attributes.areaacres &&
+          `${formatNumber(attributes.areaacres)} acres / ${formatNumber(
+            attributes.areasqkm,
+            2,
+          )} km²`,
```
```suggestion
        (attributes.areaacres
          ? `${formatNumber(attributes.areaacres)} acres`
          : '') +
          (attributes.areasqkm && attributes.areaacres ? ' / ' : '') +
          (attributes.areasqkm
            ? `${formatNumber(attributes.areasqkm, 2)} km²`
            : ''),
```

The upstream watershed popup no longer displays anything for the area. Unfortunately, that layer doesn't have the `areaacres` attribute. We should try to bring this up at tomorrow's meeting as well. Maybe the upstream watershed service could be updated to return acres. 

The above suggestion will get it back to what it was before.

We could also try converting sqkm to acres, like below.

```suggestion
        attributes.areasqkm &&
          `${formatNumber(
            attributes.areaacres
              ? attributes.areaacres
              : attributes.areasqkm / 0.00404686,
          )} acres / ${formatNumber(attributes.areasqkm, 2)} km²`,
```

maxdiebold-erg replied:
Thanks, I forgot that the square acres aren't in that service. I think calculating the value is fine.
